File: packages/core/src/presentation/presentationFormat.ts
import type { DeliveryKind } from "./types.js";

/**
 * The six presentation formats. The agent picks exactly one per item up front
 * and surfaces it to the user with a fixed, scannable header + body.
 */
export type PresentationFormat =
  | "testing"
  | "comparing"
  | "inspecting"
  | "reviewing"
  | "user_led"
  | "explaining";

export type FormatActor = "agent" | "user";

export type FormatMeta = {
  emoji: string;
  verb: string;
  descriptor: string;
  actor: FormatActor;
};

/**
 * Single source of truth for the format emoji + header copy. The emoji are
 * load-bearing and live ONLY here; every other module references this table so
 * render and selection stay consistent.
 */
export const FORMAT_META: Record<PresentationFormat, FormatMeta> = {
  testing: { emoji: "🧪", verb: "Testing", descriptor: "runs it live", actor: "agent" },
  comparing: { emoji: "⚖️", verb: "Comparing", descriptor: "guardrail / before-after", actor: "agent" },
  inspecting: { emoji: "🔎", verb: "Inspecting", descriptor: "examine the real artifact", actor: "agent" },
  reviewing: { emoji: "📜", verb: "Reviewing", descriptor: "cite an earlier run", actor: "agent" },
  user_led: { emoji: "🙋", verb: "Over to you", descriptor: "needs the human", actor: "user" },
  explaining: { emoji: "💬", verb: "Explaining", descriptor: "description only", actor: "agent" }
};

/**
 * Project a chosen format onto the legacy `delivery_kind` enum. `user_led` has
 * no native delivery_kind, so it falls back to the verification-derived base
 * kind supplied by selection.
 */
export function formatToDeliveryKind(format: PresentationFormat, baseDeliveryKind: DeliveryKind): DeliveryKind {
  switch (format) {
    case "testing":
    case "comparing":
      return "live_demo";
    case "inspecting":
    case "reviewing":
      return "evidence_review";
    case "explaining":
      return "explanation";
    case "user_led":
      return baseDeliveryKind;
  }
}

/** The default format for a verification-derived delivery kind. */
export function defaultFormatForDeliveryKind(kind: DeliveryKind): PresentationFormat {
  switch (kind) {
    case "live_demo":
      return "testing";
    case "evidence_review":
      return "reviewing";
    case "explanation":
      return "explaining";
  }
}

export type ChoosePresentationFormatOptions = {
  baseDeliveryKind: DeliveryKind;
  needsUser?: boolean;
  isContrast?: boolean;
};

/**
 * Pick a presentation format from the base delivery kind plus two affordances:
 * `needsUser` (human-in-the-loop) wins first, then `isContrast` (guardrail /
 * before-after), otherwise the default-for-delivery-kind.
 */
export function choosePresentationFormat(options: ChoosePresentationFormatOptions): PresentationFormat {
  if (options.needsUser) {
    return "user_led";
  }
  if (options.isContrast) {
    return "comparing";
  }
  return defaultFormatForDeliveryKind(options.baseDeliveryKind);
}

File: UseCasesCore/Sources/UseCasesCore/Presentation/PresentationFormat.swift
/// How a plan item is delivered, as the legacy `delivery_kind` enum spells it.
public enum DeliveryKind: String, Sendable, Equatable, CaseIterable {
  case liveDemo = "live_demo"
  case evidenceReview = "evidence_review"
  case explanation
}

/// Who acts on a card.
public enum FormatActor: String, Sendable, Equatable {
  case agent
  case user
}

/// The emoji and header copy of one format
/// (packages/core/src/presentation/presentationFormat.ts `FORMAT_META`).
public struct FormatMetadata: Sendable, Equatable {
  public let emoji: String
  public let verb: String
  public let descriptor: String
  public let actor: FormatActor
}

/// The six presentation formats. The agent picks exactly one per item up front
/// and surfaces it with a fixed header and body.
public enum PresentationFormat: String, Sendable, Equatable, CaseIterable {
  case testing
  case comparing
  case inspecting
  case reviewing
  case userLed = "user_led"
  case explaining

  /// The single source of the format emoji and header copy. The emoji are
  /// spelled as scalars so a variation selector cannot be lost in an edit:
  /// Comparing's balance scale is U+2696 followed by U+FE0F.
  public var metadata: FormatMetadata {
    switch self {
    case .testing:
      FormatMetadata(emoji: "\u{1F9EA}", verb: "Testing", descriptor: "runs it live", actor: .agent)
    case .comparing:
      FormatMetadata(
        emoji: "\u{2696}\u{FE0F}",
        verb: "Comparing",
        descriptor: "guardrail / before-after",
        actor: .agent,
      )
    case .inspecting:
      FormatMetadata(
        emoji: "\u{1F50E}",
        verb: "Inspecting",
        descriptor: "examine the real artifact",
        actor: .agent,
      )
    case .reviewing:
      FormatMetadata(
        emoji: "\u{1F4DC}",
        verb: "Reviewing",
        descriptor: "cite an earlier run",
        actor: .agent,
      )
    case .userLed:
      FormatMetadata(
        emoji: "\u{1F64B}",
        verb: "Over to you",
        descriptor: "needs the human",
        actor: .user,
      )
    case .explaining:
      FormatMetadata(
        emoji: "\u{1F4AC}",
        verb: "Explaining",
        descriptor: "description only",
        actor: .agent,
      )
    }
  }

  /// `formatToDeliveryKind`: the legacy projection. `user_led` has no delivery
  /// kind of its own, so it keeps the verification-derived base kind.
  public func deliveryKind(base: DeliveryKind) -> DeliveryKind {
    switch self {
    case .testing, .comparing: .liveDemo
    case .inspecting, .reviewing: .evidenceReview
    case .explaining: .explanation
    case .userLed: base
    }
  }

  /// `defaultFormatForDeliveryKind`.
  public static func defaultFormat(for kind: DeliveryKind) -> PresentationFormat {
    switch kind {
    case .liveDemo: .testing
    case .evidenceReview: .reviewing
    case .explanation: .explaining
    }
  }

  /// `choosePresentationFormat`: a human actor wins, then a contrast, then the
  /// default for the base delivery kind.
  public static func choose(
    baseDeliveryKind: DeliveryKind,
    needsUser: Bool,
    isContrast: Bool,
  ) -> PresentationFormat {
    if needsUser {
      return .userLed
    }
    if isContrast {
      return .comparing
    }
    return defaultFormat(for: baseDeliveryKind)
  }
}
